[GHF] Add revert workflow

malfet opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #71868
* #71867

This adds `try_revert` repository dispatch that will revert commit
that were previosly landed by merge workflow if requested by org member

malfet committed:
[GHF] Add revert workflow

This adds `try_revert` repository dispatch that will revert commit
that were previosly landed by merge workflow if requested by org member

[ghstack-poisoned]

malfet committed:
Update on "[GHF] Add revert workflow"


This adds `try_revert` repository dispatch that will revert commit
that were previosly landed by merge workflow if requested by org member

[ghstack-poisoned]

bigfootjon commented on .github/scripts/trymerge.py:
```diff
@@ -310,16 +353,58 @@ def check_if_should_be_merged(pr: GitHubPR, repo: GitRepo) -> None:
+    if commit_sha is None:
+        commits = repo.commits_resolving_gh_pr(pr.pr_num)
+        if len(commits) == 0:
+            raise RuntimeError("Can't find any commits resolving PR")
```
Just verifying, this should always be added by the `trymerge` workflow? So this error message should be "impossible" at runtime?

malfet replied:
I didn't find a query to verify, that PR was closed because commit resolving it were landed to master, vs user just closed it for no reason. So this check is always necessary

bigfootjon commented on .github/scripts/trymerge.py:
```diff
@@ -310,16 +353,58 @@ def check_if_should_be_merged(pr: GitHubPR, repo: GitRepo) -> None:
+    if author_association != "MEMBER":
+        return post_comment(f"Will not revert as @{author_login} is not a member, but {author_association}")
+
+    find_matching_merge_rule(pr, repo)
```
Maybe add a comment that this raises on error

malfet replied:
I have it on the function description, but you are right, it would not hurt here

bigfootjon commented on .github/scripts/trymerge.py:
```diff
@@ -310,16 +353,58 @@ def check_if_should_be_merged(pr: GitHubPR, repo: GitRepo) -> None:
+        gh_post_comment(pr.org, pr.project, pr.pr_num, msg, dry_run=dry_run)
+    if not pr.is_closed():
+        return post_comment(f"Can't revert open PR #{pr.pr_num}")
+    if not RE_REVERT_CMD.match(pr.get_comment_body()):
```
What if someone posts a comment in between the revert command and the workflow getting to this point?

Shouldn't there be a search for the last N comments to find the revert command regex, then pass that offset to the below functions?

malfet replied:
I've considered that, but this adds complexity to the code where it is not, strictly speaking needed. If somebody does that, that author can try again. Or, we can ask probot to pass a comment id as an option. Will do as followup

malfet committed:
Update on "[GHF] Add revert workflow"


This adds `try_revert` repository dispatch that will revert commit
that were previosly landed by merge workflow if requested by org member

[ghstack-poisoned]

bigfootjon reviewed (approved):
I'm tentatively ok with this, we can follow up with another PR for my concern about race condition with comments if it becomes a problem

malfet committed:
Update on "[GHF] Add revert workflow"


This adds `try_revert` repository dispatch that will revert commit
that were previosly landed by merge workflow if requested by org member

[ghstack-poisoned]

malfet committed:
Update on "[GHF] Add revert workflow"


This adds `try_revert` repository dispatch that will revert commit
that were previosly landed by merge workflow if requested by org member

[ghstack-poisoned]

malfet committed:
Update on "[GHF] Add revert workflow"


This adds `try_revert` repository dispatch that will revert commit
that were previosly landed by merge workflow if requested by org member

[ghstack-poisoned]

seemethere commented:
@pytorchbot merge this

malfet commented:
@pytorchbot merge this

malfet commented:
> Merge failed due to Command '['git', '-C', '/home/runner/work/pytorch/pytorch', 'push', 'origin', 'master']' returned non-zero exit status 1.

See https://github.com/pytorch/pytorch/actions/runs/1753260110 failed due to the race condition, probably good followup items are:
 - Add references to runlog and push errors to the message
 -  Auto-retry  pushes on races
